x/mongo/driver/topology/pool.go

```diff
@@ -601,3 +527,392 @@ func (p *pool) clear(serviceID *primitive.ObjectID) {
+// A wantConn records state about a wanted connection (that is, an active call to checkOut).
+// The conn may be gotten by creating a new connection or by finding an idle connection, or a
+// cancellation may make the conn no longer wanted. These three options are racing against each
+// other and use wantConn to coordinate and agree about the winning outcome.
+// Based on https://cs.opensource.google/go/go/+/refs/tags/go1.16.6:src/net/http/transport.go;l=1174-1240
+type wantConn struct {
+	ready chan struct{}
+
+	mu   sync.Mutex // Guards conn, err
+	conn *connection
+	err  error
+}
+
+func newWantConn() *wantConn {
+	return &wantConn{
+		ready: make(chan struct{}, 1),
+	}
+}
+
+// waiting reports whether w is still waiting for an answer (connection or error).
+func (w *wantConn) waiting() bool {
+	select {
+	case <-w.ready:
+		return false
+	default:
+		return true
+	}
+}
+
+// tryDeliver attempts to deliver conn, err to w and reports whether it succeeded.
+func (w *wantConn) tryDeliver(conn *connection, err error) bool {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.conn != nil || w.err != nil {
+		return false
+	}
+
+	w.conn = conn
+	w.err = err
+	if w.conn == nil && w.err == nil {
+		panic("x/mongo/driver/topology: internal error: misuse of tryDeliver")
+	}
+	close(w.ready)
+	return true
+}
+
+// cancel marks w as no longer wanting a result (for example, due to cancellation). If a connection
+// has been delivered already, cancel returns it with p.checkInNoEvent(). Note that the caller must
+// not hold any locks on the pool while calling cancel.
+func (w *wantConn) cancel(p *pool, err error) {
+	w.mu.Lock()
+	if w.conn == nil && w.err == nil {
+		close(w.ready) // catch misbehavior in future delivery
+	}
+	conn := w.conn
+	w.conn = nil
+	w.err = err
+	w.mu.Unlock()
+
+	if conn != nil {
+		_ = p.checkInNoEvent(conn)
+	}
+}
+
+// A wantConnQueue is a queue of wantConns.
+// Based on https://cs.opensource.google/go/go/+/refs/tags/go1.16.6:src/net/http/transport.go;l=1242-1306
+type wantConnQueue struct {
+	// This is a queue, not a deque.
+	// It is split into two stages - head[headPos:] and tail.
+	// popFront is trivial (headPos++) on the first stage, and
+	// pushBack is trivial (append) on the second stage.
+	// If the first stage is empty, popFront can swap the
+	// first and second stages to remedy the situation.
+	//
+	// This two-stage split is analogous to the use of two lists
+	// in Okasaki's purely functional queue but without the
+	// overhead of reversing the list when swapping stages.
+	head    []*wantConn
+	headPos int
+	tail    []*wantConn
+}
+
+// len returns the number of items in the queue.
+func (q *wantConnQueue) len() int {
+	return len(q.head) - q.headPos + len(q.tail)
+}
+
+// pushBack adds w to the back of the queue.
+func (q *wantConnQueue) pushBack(w *wantConn) {
+	q.tail = append(q.tail, w)
+}
+
+// popFront removes and returns the wantConn at the front of the queue.
+func (q *wantConnQueue) popFront() *wantConn {
```

Comment:
Question: is returning `nil` ergonomic? Seems easy to forget to check `w == nil` in the caller and end up with a nil-pointer dereference. Would it be slightly better to return `*wantConn, bool` or something like that instead?

Reply:
As far as readability, I do think `w, ok := q.popFront()` is actually more readable. However, getting rid of the `nil` check also depends on getting the implementation of `popFront()` and `pushBack()` correct. If there are any bugs, then you could potentially get back `nil, true`, which would potentially cause a panic if you're not also checking `w == nil`. If you do check `w == nil` then that will be a non-impacting bug, but then the `true` becomes redundant.

tldr: I don't feel that the improved readability is worth the additional possibility for unchecked nil pointers.